panel/panels.py

```diff
@@ -327,32 +327,57 @@ def _update(self, model):
         div.update(**self._get_properties())
 
 
-class MatplotlibPanel(DivBasePanel):
+class PNGPanel(DivBasePanel):
     """
-    A MatplotlibPanel renders a matplotlib figure to png and wraps
-    the base64 encoded data in a bokeh Div model.
+    Encodes a PNG as base64 and wraps it in a Bokeh Div model.
+    This base class supports anything with a _repr_png_ method, but
+    subclasses can provide their own way of obtaining or generating a PNG.
     """
 
     @classmethod
     def applies(cls, obj):
-        return obj.__class__.__name__ == 'Figure' and hasattr(obj, '_cachedRenderer')
-
+        return hasattr(obj, '_repr_png_')
+
+    def _png(self):
+        return self.object._repr_png_()
+    
+    def _pngshape(self, data):
+        """Calculate and return PNG width,height"""
+        import struct
+        w, h = struct.unpack('>LL', data[16:24])
+        return int(w), int(h)
+    
     def _get_properties(self):
-        bytes_io = BytesIO()
-        self.object.canvas.print_figure(bytes_io)
-        data = bytes_io.getvalue()
+        data = self._png()
         b64 = base64.b64encode(data).decode("utf-8")
         src = "data:image/png;base64,{b64}".format(b64=b64)
-        width, height = self.object.canvas.get_width_height()
         html = "<img src='{src}'></img>".format(src=src)
         
-        p = super(MatplotlibPanel,self)._get_properties()
+        p = super(PNGPanel,self)._get_properties()
+        width, height = self._pngshape(data)
         if self.width  is None: p["width"]  = width
         if self.height is None: p["height"] = height
```

Comment:
This should be:

```python
p['width'] = self.width or width
p['height'] = self.height or height
```

The width and height should always be provided.

Reply:
Nevermind it's fine I see the superclass defines it.

Reply:
Shouldn't it be possible for the user to set self.width or self.height to zero, if they want to?  The proposed replacement code doesn't allow that.

Reply:
(Since self.width and self.height are only for user overriding of the computed width,height).

Reply:
Oh, I see; you mean that `p["width"]` and `p["height"]` should always be provided.  So I guess it should be:

```
p["width"]  = width  if self.width  is None else self.width
p["height"] = height if self.height is None else self.height
```

Reply:
Oops; I never saw your "nevermind"; somehow the PR page updated with only your first comment.